### PR TITLE
Initial functionality

### PR DESCRIPTION
**Pull request by shadyvb**
This presents the initial functionality, with basic unit tests for the basic functionality of the transformer class.

Next steps:

- [x] Add more unit tests for the SegmentTransformer class
- [x] Add readme with information about public API and transformation maps data structure
- [x] Add proper integration tests with real dummified data and remove the dummy bootstrap.php file

ref https://github.com/humanmade/altis-analytics/issues/220

### CODE REVIEW

File: composer.json
Comment (roborourke) [inline composer.json]:
```suggestion
    "license": "GPL-2.0-or-later",
```

File: inc/class-segment-transformer.php
Comment (roborourke) [inline inc/class-segment-transformer.php]: Technically if used they should exist at `endpoint.User.UserAttributes`.

File: inc/class-segment-transformer.php
Comment (roborourke) [inline inc/class-segment-transformer.php]: Can add this one in too, not sure why it's omitted here so apologies if I'm missing something.
```suggestion
					'content' => 'attributes.qv_utm_content',
```

File: inc/class-segment-transformer.php
Comment (roborourke) [inline inc/class-segment-transformer.php]: Why `endpoint.Attributes` here? I think this should be `attributes` and also `metrics` as those are the event level custom data points. Not sure how well Segment deals with nested data, `attributes` and `metrics` could technically be merged here but that would complicate the transform mapping / system I think.

File: inc/class-segment-transformer.php
Comment (roborourke) [inline inc/class-segment-transformer.php]: Need to confirm these are unique per event so I'd check a few. I have a feeling we shouldn't use this value after all as it's intended to only reflect when a change was made to the endpoint data. I'm going to dig into that a bit further but we may just need to set this as a hash of the event data for Segment's purposes.

File: inc/class-transformer.php
Comment (roborourke) [inline inc/class-transformer.php]: This sort of seems like what happens by default, I'm a bit unsure as to why it's returning the path component though and not the value. Should it just be using the `$key` when looping over the maps?

File: inc/export/namespace.php
Comment (roborourke) [inline inc/export/namespace.php]: What sort of use cases are you seeing for these hooks?

File: inc/export/namespace.php
Comment (roborourke) [inline inc/export/namespace.php]: Looks like this is used in another place so might be useful to make a getter function so there's a single instance of the filter.

File: inc/export/namespace.php
Comment (roborourke) [inline inc/export/namespace.php]: Note that `E_USER_ERROR` will cause PHP to exit as if there was a fatal error so the following action won't be triggered.

File: inc/namespace.php
Comment (roborourke) [inline inc/namespace.php:16]: This is quite a minimal file so I'm not sure there's much benefit to having the `Export` sub namespace. Not required of course just something to consider unless you think there might be more to add here in future.

Comment (roborourke) [reply inc/namespace.php:16]: Ah is it to make it easier to write the tests?

File: inc/export/namespace.php
Comment (roborourke) [inline inc/export/namespace.php]: Are these intended to be definable as global constants? Is the Batch URL changeable? Thought it was a fixed value https://api.segment.io/v1/batch

File: inc/class-transformer.php
Comment (roborourke) [inline inc/class-transformer.php]: I was wondering why these classes are under the `Export` namespace but not the directory. Is the `Export` part of the namespace necessary?

File: inc/class-segment-transformer.php
Comment (roborourke) [inline inc/class-segment-transformer.php]: Feels like a switch might be more suitable here, but also could these customisations be done in the specific methods e.g. `track()`, `page()` etc...?

File: composer.json
Comment (roborourke) [inline composer.json]: Don't need to specify these if using classmap

File: plugin.php
Comment (roborourke) [inline plugin.php]: I think these are redundant if you're using the `autoload` functionality in `composer.json`.

**Review by roborourke — COMMENTED** [review]
Interesting approach. At the moment it seems like the `Transform` class could be a generic helper if it did more of the processing for ndjson and perhaps required extending it with an event mapping method. Hard to say with this being the 1st integration though. The way its used currently and the very Segment specific subclass I think it doesn't necessarily need to be an object, there's only a single instance used and it doesn't seem to need or use the constructor argument right now from the looks of it. This package is specific to the Segment integration so I'd consider simplifying it for the time being.

File: inc/class-segment-transformer.php
Comment (shadyvb) [reply inc/class-segment-transformer.php]: We can merge trees alright now, that's what happens when you add a numeric index key like `endpoint.Attributes` already.

~It's the discrepancy in logged data that throws me off. I was trying to examine existing data and match to that then confirm my assumptions with you. Let's review some data and discuss on Slack.~
nvm, this is the event-level attributes, you're totally right.

File: inc/class-segment-transformer.php
Comment (shadyvb) [reply inc/class-segment-transformer.php]: `RequestId` does not seem unique from data, also the batch API endpoint doesn't really show `messageId` within examples nor in docs, so I'll assume this isn't required. Typically Segment SDKs create this to be able to pinpoint specific messages. So I removed the key altogether, let me know if I should not.

File: inc/class-transformer.php
Comment (shadyvb) [reply inc/class-transformer.php]: The `$path` in this instance when using something like `foobar|static` is `foobar`, and I use that as a way to let the parser know they should just use the string and not try to map it from the event data. eg: `type: 'identify|static'`.

File: inc/export/namespace.php
Comment (shadyvb) [reply inc/export/namespace.php]: Good question, can't think of reasons OTTOMH. Do you think any of them would be valuable to have ?

File: inc/export/namespace.php
Comment (shadyvb) [reply inc/export/namespace.php]: Of course :facepalm:

Comment (shadyvb) [reply inc/export/namespace.php]: Switched them to warnings instead, as we'd need to log all of them rather than just the first one.

File: inc/namespace.php
Comment (shadyvb) [reply inc/namespace.php:16]: Not really, I do think we might extend this in the future to more than just batch backfiling of events. So gave it room to grow.

File: inc/export/namespace.php
Comment (shadyvb) [reply inc/export/namespace.php]: The URL should not be changeable, no. I just needed to define a local constant for it for visibility.

Re the key, we just need it to be defined in a global const, right ? we don't even need to filter it. Too many filters :facepalm:

File: inc/class-transformer.php
Comment (shadyvb) [reply inc/class-transformer.php]: They should be moved, yes. Did so now.

File: inc/class-segment-transformer.php
Comment (shadyvb) [reply inc/class-segment-transformer.php]: A `switch` might be more readable 👍 , will go for that. I preferred to encapsulate all the mapping definition in one function so any changes are more readable / contextual.

**Comment by shadyvb** [comment]
> At the moment it seems like the Transform class could be a generic helper if it did more of the processing for ndjson and perhaps required extending it with an event mapping method. 

`Transformer` was meant to be a generic one, that maps data from one format to another. There was a few similar open-source classes available but none did the job as needed or was simple enough to fork / optimize. Its single purpose is to transform data, so I'd rather keep it as simple as that.

`SegmentTransformer` is then meant to extend the transformer with custom mapping logic ( not just data ), eg: see how groups work. 

> The way its used currently and the very Segment specific subclass I think it doesn't necessarily need to be an object, there's only a single instance used and it doesn't seem to need or use the constructor argument right now from the looks of it. This package is specific to the Segment integration so I'd consider simplifying it for the time being.

While Segment doesn't need constructor arguments, main reason is to be able to construct an object once, do all the necessary processing of maps, and have that be available to process large amounts of data without having to execute redundant steps. Specially given the volume of data we'll need to parse all the time.

`Transformer` can still be used with a simple single map, I can imagine that working for other exporters as well. Hence the abstraction. I did not want to package it separately for now for the sake of time/effort, but it can be moved in the future.

The abstraction made it easy to test it as well, which is a big bonus.

File: inc/class-segment-transformer.php
Comment (shadyvb) [reply inc/class-segment-transformer.php]: Updated after this was fixed in event data in https://github.com/humanmade/aws-analytics/pull/353/files

File: inc/export/namespace.php
Comment (shadyvb) [reply inc/export/namespace.php]: Removed.

File: composer.json
Comment (roborourke) [inline composer.json]: No need to repeat "altis" if we're in that namespace

File: README.md
Comment (roborourke) [inline README.md]: Make sure this matches the name in `composer.json`

File: README.md
Comment (roborourke) [inline README.md]: I guess this could work as a standalone plugin, but for Altis the platform we don't want folks modifying `wp-config.php` directly. Are you thinking this package could be published and installed separately as a plugin for Accelerate plugin users?

That plugin already bundles the analytics module effectively so for official integrations we probably don't need to worry about them being standalone.

File: README.md
Comment (roborourke) [inline README.md]:
```suggestion
**`altis.analytics.segment.api_write_key : String`**
```

File: README.md
Comment (roborourke) [inline README.md]: In real terms this would come from `endpoint.Location.Country`.

File: README.md
Comment (roborourke) [inline README.md]: Not sure it's clear with the namespace being used that this is pipe delimited, for the sake of the example it could just be a simple sanitisation function call like `floatval` or similar

File: composer.json
Comment (roborourke) [inline composer.json:5]: Missing package type definition here? Seems like it should be defined as a `wordpress-plugin`

File: plugin.php
Comment (roborourke) [inline plugin.php:20]: There are only dev requirements, so I think you could use this approach and not bother with the requires below or just require all the files including the classes and leave this out.

This is a sort of hybrid approach that doesn't really make sense at the moment as it's only a composer installed copy, or a built version with a vendor directory that is going to be able to include all the required files.

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php]:
```suggestion
					'version' => 'endpoint.Demographic.AppVersion',
```

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php]: May need an alternative approach to this e.g. `attributes.x-amz-request-id` or fallback to hash of event JSON.

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php]:
```suggestion
					'endpoint.User.UserAttributes',
```

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php]: How reliable is this? I know for all stored timestamps in the Altis code its milliseconds. This feels like a bit of a dangerous assumption to me so it'd be better to have 2 separate functions IMO, one for seconds and one for millis so it's obvious what you're getting.

Not sure if we even need one for seconds tbh.

File: inc/export/class-transformer.php
Comment (roborourke) [inline inc/export/class-transformer.php]: Should probably check that `$callback` `is_callable()` here.

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php:111]: How are arrays handled in this situation? All `Attributes` and `UserAttributes` under the endpoint are arrays of values, can segment handle that? From the tests it looks like they're converted to strings and maybe just the first item in an array is included? We should pass all values through as an array to Segment if possible.

File: inc/export/class-segment-transformer.php
Comment (roborourke) [inline inc/export/class-segment-transformer.php:110]: Realised we opted to remove the specifically gathered pageViews and sessions metrics that were included here previously but perhaps we should include `endpoint.Metrics` as a whole after all, as there might be custom project specific values in there. What do you think?

**Review by roborourke — DISMISSED** [review]
For a package that is intended to be a dependency you should leave out the `composer.lock` file. It's only there for development purposes really so doesn't need to be maintained in the same way as a root project's `composer.lock`.

Couple of things to change and some questions it'd be good to confirm before we move ahead with this.

File: composer.json
Comment (shadyvb) [reply composer.json]: Hmmm, I'd think `altis-analytics` is a module name rather than just belonging to Altis. Would keep it as so unless you feel strongly about it, feels a bit weird without it.

File: README.md
Comment (shadyvb) [reply README.md]: Nice catch!

File: inc/export/class-segment-transformer.php
Comment (shadyvb) [reply inc/export/class-segment-transformer.php]: Going for a hash of the event JSON object to avoid conflicts and to be backward compat given the recent changes in `RequestId` input.

File: inc/export/class-segment-transformer.php
Comment (shadyvb) [reply inc/export/class-segment-transformer.php:111]: I've asked their support on https://segment.zendesk.com/hc/en-us/requests/466909 and we'll see what they respond with.

File: inc/export/class-segment-transformer.php
Comment (shadyvb) [reply inc/export/class-segment-transformer.php:110]: I'm still concerned about batch uploading (in parallel requests) different values for these metrics, but perhaps Segment would be smart enough to understand the precedence based on timestamps.
In a perfect world, Segment should deduce such metrics based on collected events, both sessions and pageViews should be aggregates of how many sessions are created and how many page view track events have been logged. But I'm unsure how the reporting works in Segment yet, so might not be so easy to deduce.
Let's log them and see what eventually gets displayed.

File: composer.json
Comment (roborourke) [reply composer.json]: `altis/analytics` is the module's package name so I'd still opt to not duplicate the `altis` component personally but I'll leave the final decision up to you.

**Review by roborourke — APPROVED** [review]
Looking good - pre-approving, there's just the questions you're waiting on from the looks of it and a decision to make on the package name but otherwise good to go here. Nice work.

File: composer.json
Comment (shadyvb) [reply composer.json]: Makes sense, updated.

**Pull request merged**
